src/driver/compile.c

```diff
@@ -193,55 +362,68 @@ static int yf_compile_project(struct yf_args * args) {
         }
         return 0;
     }
-    
-    ret = yf_run_compiler_on_data(&data, args);
-
-    yf_cleanup(&data);
 
-    return ret;
+    return yf_create_compiler_jobs(compilation, &data, args);
 
 }
 
-static int yf_compile_files(struct yf_args * args) {
+static int yf_compile_files(struct yf_args * args, struct yf_compilation_data * compilation) {
     
     struct yf_project_compilation_data data;
-    int i, ret;
-    struct yf_file_compilation_data * fdata;
+    struct yf_compilation_unit_info * fdata;
+    int i;
 
     /* No project name */
-    data.project_name = yf_malloc(50);
-    strcpy(data.project_name, "");
+    data.project_name = NULL;
 
     data.files = yfh_new();
 
     for (i = 0; i < args->num_files; ++i) {
-        fdata = malloc(sizeof(struct yf_file_compilation_data));
-        memset(fdata, 0, sizeof (struct yf_file_compilation_data));
-        fdata->file_name = args->files[i];
+        fdata = malloc(sizeof(struct yf_compilation_unit_info));
+        memset(fdata, 0, sizeof (struct yf_compilation_unit_info));
+        fdata->file_name = yf_strdup(args->files[i]);
         fdata->parse_anew = 1;
         /* TODO - more data */
         yfh_set(data.files, fdata->file_name, fdata);
     }
 
-    ret = yf_run_compiler_on_data(&data, args);
-    yf_cleanup(&data);
-    return ret;
+    return yf_create_compiler_jobs(compilation, &data, args);
+
+}
+
+static int yfc_compile(
+    struct yf_compilation_data * pdata,
+    struct yf_compile_compile_job * udata
+) {
+
+    struct yf_compile_analyse_job * adata = udata->unit;
+    int retval;
+
+    retval = yf_validate_ast(pdata, adata);
+    if (retval)
+        return retval;
+
+    if (adata->stage >= YF_COMPILE_CODEGEN)
+        retval = yf_backend_generate_code(adata);
+
+    return retval;
 
 }
 
 /**
  * Run the lexing and parsing on one file.
  */
-static int yf_run_frontend(
-    struct yf_file_compilation_data * file,
-    struct yf_args * args
+static int yfc_analyse(
```

Comment:
I think `yfc_frontend` or similar might be a more accurate name because "analysis" could refer to lexical analysis, syntactic analysis, or semantic analysis.

Reply:
Changed